typescript/schemas/src/internal/generateRust.ts

```diff
@@ -0,0 +1,236 @@
+    (schema) => {
+      const { fields, description } = schema;
+      const name = schema.name.replace("JSON", "Json");
+      const snakeName = toSnakeCase(name);
+      return`\
+${formatComment(description)}
+#[repr(C)]
+pub struct ${name} {
+  ${fields
+    .flatMap((field) => {
+      const comment = formatComment(field.description);
+      const identName = escapeId(toSnakeCase(field.name));
+      let fieldType: string;
+      let fieldHasLen = false;
+      switch (field.type.type) {
+        case "primitive":
+          if (field.type.name === "bytes") {
+            fieldType = "*const c_uchar";
+            fieldHasLen = true;
+          } else if (field.type.name === "time") {
+            fieldType = "*const Timestamp";
+          } else if (field.type.name === "duration") {
+            fieldType = "*const Duration";
+          } else {
+            fieldType = primitiveToRust(field.type.name);
+          }
+          break;
+        case "enum":
+          fieldType = `Foxglove${field.type.enum.name}`;
+          break;
+        case "nested":
+          fieldType = `*const ${field.type.schema.name.replace("JSON", "Json")}`;
+          break;
+      }
+      const lines: string[] = [comment];
+      if (typeof field.array === "number") {
+        lines.push(`pub ${identName}: [${fieldType}; ${field.array}],`);
+        if (fieldHasLen) {
+          lines.push(`pub ${identName}_len: [usize; ${field.array}],`);
+        }
+      } else if (field.array === true) {
+        lines.push(`pub ${identName}: *const ${fieldType},`);
+        if (fieldHasLen) {
+          lines.push(`pub ${identName}_len: *const usize,`);
+        }
+        lines.push(`pub ${identName}_count: usize,`);
+      } else {
+        lines.push(`pub ${identName}: ${fieldType},`);
+        if (fieldHasLen) {
+          lines.push(`pub ${identName}_len: usize,`);
+        }
+      }
+      return lines.join("\n");
+    })
+    .join("\n\n")}
+}
+
+impl ${name} {
+${name.endsWith("Primitive") ? "" : `
+  unsafe fn borrow_option_to_native(msg: Option<&Self>) -> Result<ManuallyDrop<foxglove::schemas::${name}>, foxglove::FoxgloveError> {
+    let Some(msg) = msg else {
+      return Err(foxglove::FoxgloveError::ValueError("msg is required".to_string()));
+    };
+    unsafe { msg.borrow_to_native() }
+  }
+`}
+
+  unsafe fn borrow_to_native(&self) -> Result<ManuallyDrop<foxglove::schemas::${name}>, foxglove::FoxgloveError> {
+    Ok(ManuallyDrop::new(foxglove::schemas::${name} {
+    ${fields
+      .map((field) => {
+        const srcName = escapeId(toSnakeCase(field.name));
+        const dstName = escapeId(toSnakeCase(field.name));
+        if (field.array != undefined) {
+          if (typeof field.array === "number") {
+            return `${dstName}: unsafe { Vec::from_raw_parts(self.${srcName}.as_ptr() as *mut ${(field.type.type === "primitive") ? primitiveToRust(field.type.name) : "todo!()"}, self.${srcName}.len(), self.${srcName}.len()) }`;
```

Comment:
can we replace the todo! with a TS assertion/error? Looks like we're not generating this (yet?).

Reply:
Yeah, I swapped it out for throwing an error, haven;t updated this branch yet